test/Test/TreasuryDAO.hs

```diff
@@ -0,0 +1,194 @@
+-- SPDX-FileCopyrightText: 2020 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module Test.TreasuryDAO
+  ( test_TreasuryDAO
+  ) where
+
+import Universum hiding (compare, drop, (>>))
+
+import Lorentz
+import Morley.Nettest
+import Morley.Nettest.Tasty
+import Test.Tasty (TestTree, testGroup)
+import Time (sec)
+
+import qualified Lorentz.Contracts.BaseDAO.Types as DAO
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Contracts.TreasuryDAO
+import Lorentz.Contracts.TreasuryDAO.Types
+import Test.Common
+
+{-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
+
+-- | Testing a Treasury-like DAO. Ex. DNS Treasury
+test_TreasuryDAO :: TestTree
+test_TreasuryDAO = testGroup "TreasuryDAO Tests"
+  [ testGroup "Proposal creator:"
+      [ nettestScenario "can propose a valid proposal" validProposal
+      , nettestScenario "can flush a 'Token transfer' proposal" flushTokenTransfer
+      , nettestScenario "can flush a 'Xtz transfer' proposal" flushXtzTransfer
+      -- TODO: flush a config proposal
+      -- TODO: check all fail scenario of proposal_check
+      ]
+  ]
+
+validProposal :: (Monad m) => NettestImpl m -> m ()
+validProposal = uncapsNettest $ do
+  ((owner1, _), (owner2, _), dao, _) <-
+      originateBaseDaoWithBalance def config
+          (\owner1_ owner2_ ->
+              [ ((owner1_, DAO.unfrozenTokenId), 200)
+              , ((owner2_, DAO.unfrozenTokenId), 100)
+              ]
+          )
+
+  let proposal = tokenTransferProposalMetadata (toAddress dao) owner2 owner1 -- transfer token from owner2 -> owner1
+      params t = DAO.ProposeParams
+        { ppFrozenToken = t
+        , ppProposalMetadata = proposal
+        }
+      expectedToken = fromInteger $ toInteger $ length $ lPackValueRaw proposal
+
+  callFrom (AddressResolved owner1) dao (Call @"Propose") (params $ expectedToken - 1)
+    & expectCustomError_ #fAIL_PROPOSAL_CHECK
+
+  callFrom (AddressResolved owner1) dao (Call @"Propose") (params $ expectedToken + 1)
+    & expectCustomError_ #fAIL_PROPOSAL_CHECK
+
+  -- Expected token is 58 in this case
+  _ <- createSampleProposal expectedToken proposal owner1 dao
+
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 117
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner1 83
+
+flushTokenTransfer :: (Monad m) => NettestImpl m -> m ()
+flushTokenTransfer = uncapsNettest $ do
+  ((owner1, _), (owner2, _), dao, admin) <-
+      originateBaseDaoWithBalance def config
+          (\owner1_ owner2_ ->
+              [ ((owner1_, DAO.unfrozenTokenId), 200)
+              , ((owner2_, DAO.unfrozenTokenId), 100)
+              ]
+          )
+  callFrom (AddressResolved admin) dao (Call @"Set_voting_period") 20
+  callFrom (AddressResolved admin) dao (Call @"Set_quorum_threshold") 1
```

Comment:
Btw, I noticed this pattern that we originate with a certain config/storage and then call these two entrypoints to update it. Can't we originate the contract with the right voting period and threshold from the beginning to make less calls? Our tests are already quite slow, so we should avoid making unnecessary calls when possible.

Reply:
I've updated this now. The PR is mostly ready for review. The only problem is that, we hit the operation size limit again: `oversized_operation\",\"size\":16933,\"max_size\":16384`. 😔